engine/src/main/java/org/pentaho/di/trans/step/BaseStep.java

```diff
@@ -0,0 +1,4415 @@
+
+    // call all row listeners...
+    for ( RowListener listener : rowListeners ) {
+      listener.errorRowWrittenEvent( rowMeta, row );
+    }
+
+    if ( errorRowSet != null ) {
+      while ( !errorRowSet.putRow( errorRowMeta, errorRowData ) ) {
+        if ( isStopped() ) {
+          break;
+        }
+      }
+      incrementLinesRejected();
+    }
+
+    verifyRejectionRates();
+  }
+
+  /**
+   * Verify rejection rates.
+   */
+  private void verifyRejectionRates() {
+    StepErrorMeta stepErrorMeta = stepMeta.getStepErrorMeta();
+    if ( stepErrorMeta == null ) {
+      return; // nothing to verify.
+    }
+
+    // Was this one error too much?
+    if ( maxErrors > 0 && getLinesRejected() > maxErrors ) {
+      logError( BaseMessages.getString( PKG, "BaseStep.Log.TooManyRejectedRows", Long.toString( maxErrors ), Long
+        .toString( getLinesRejected() ) ) );
+      setErrors( 1L );
+      stopAll();
+    }
+
+    if ( maxPercentErrors > 0
+      && getLinesRejected() > 0
+      && ( minRowsForMaxErrorPercent <= 0 || getLinesRead() >= minRowsForMaxErrorPercent ) ) {
+      int pct =
+        (int) Math.ceil( 100 * (double) getLinesRejected() / getLinesRead() ); // additional conversion for PDI-10210
+      if ( pct > maxPercentErrors ) {
+        logError( BaseMessages.getString(
+          PKG, "BaseStep.Log.MaxPercentageRejectedReached", Integer.toString( pct ), Long
+            .toString( getLinesRejected() ), Long.toString( getLinesRead() ) ) );
+        setErrors( 1L );
+        stopAll();
+      }
+    }
+  }
+
+  /**
+   * Current input stream.
+   *
+   * @return the row set
+   */
+  @VisibleForTesting
+  RowSet currentInputStream() {
+    inputRowSetsLock.readLock().lock();
+    try {
+      return inputRowSets.get( currentInputRowSetNr );
+    } finally {
+      inputRowSetsLock.readLock().unlock();
+    }
+  }
+
+  /**
+   * Find the next not-finished input-stream... in_handling says which one...
+   */
+  private void nextInputStream() {
+    blockPointer = 0;
+
+    int streams = inputRowSets.size();
+
+    // No more streams left: exit!
+    if ( streams == 0 ) {
+      return;
+    }
+
+    // Just the one rowSet (common case)
+    if ( streams == 1 ) {
+      currentInputRowSetNr = 0;
+    }
+
+    // If we have some left: take the next!
+    currentInputRowSetNr++;
+    if ( currentInputRowSetNr >= streams ) {
+      currentInputRowSetNr = 0;
+    }
+  }
+
+  /**
+   * Wait until the transformation is completely running and all threads have been started.
+   */
+  protected void waitUntilTransformationIsStarted() {
+    // Have all threads started?
+    // Are we running yet? If not, wait a bit until all threads have been
+    // started.
+    //
+    if ( this.checkTransRunning == false ) {
+      while ( !trans.isRunning() && !stopped.get() ) {
+        try {
+          Thread.sleep( 1 );
+        } catch ( InterruptedException e ) {
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**: InterruptedException is ignored. This can delay thread shutdown and clear the thread’s interrupt status. Only code that implements a thread’s interruption policy can swallow an interruption request.

**Fix**: Rethrow the InterruptedException or reinterrupt the current thread using *Thread.currentThread().interrupt()* so that higher-level interrupt handlers can function correctly.
 If you are wrapping the InterruptedException inside a RuntimeException, call *Thread.currentThread().interrupt()* before throwing the RuntimeException.

Learn more about [interrupts](https://docs.oracle.com/javase/tutorial/essential/concurrency/interrupt.html) and [dealing with InterruptedException](https://www.ibm.com/developerworks/java/library/j-jtp05236/index.html?ca=drs-#N10187)